homeassistant/helpers/entity.py

```diff
@@ -281,6 +261,32 @@ def async_set_context(self, context):
                             "https://goo.gl/Nvioub", self.entity_id,
                             type(self), end - start)
 
+        return (state, attr)
+
+    async def async_update_ha_state(self, force_refresh=False, stateattr=None):
+        """Update Home Assistant with current state of entity.
+
+        If force_refresh == True will update entity before setting state.
+
+        This method must be run in the event loop.
+        """
+        if self.hass is None:
+            raise RuntimeError("Attribute hass is None for {}".format(self))
+
+        if self.entity_id is None:
+            raise NoEntitySpecifiedError(
+                "No entity id specified for entity {}".format(self.name))
+
+        # update entity data
+        if force_refresh:
+            try:
+                await self.async_device_update()
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Update for %s fails", self.entity_id)
+                return
+
+        state, attr = stateattr or self._get_state_and_attributes()
```

Comment:
I know that `stateattr` is passed in as `None` when force refresh, but it is still confusing to read the code as such. Please put an explicit `state, attr = self._get_state_and_attributes()` inside the `if force_refresh`. You probably also want to safe guard and check if `stateattr is None` too. I know you do all that now, but it's hard to follow and this is important code for people to understand.